src/include/denc.h

```diff
@@ -307,6 +349,14 @@ inline void denc_signed_varint(T& v, bufferptr::iterator& p)
 inline void denc_varint_lowz(uint64_t v, size_t& p) {
   p += sizeof(v) + 2;
 }
+
+inline void denc_varint_lowz_exact(uint64_t v, size_t& p) {
+  int lowznib = v ? (ctz(v) / 4) : 0;
+  if (lowznib > 3)
+    lowznib = 3;
+  v >>= lowznib * 4 - 2;
```

Comment:
This would make perfect sense, but only if we would be working on perfectly predictable processors -- these are long gone. Processors are currently free to prefetch, decode and reorder instructions to fit available ALUs and other resources (read, for example, about instruction fusing), and compilers do the same. In that particular case, compiler can translate this to just one shift (instead of two or shift and multiply) and subtraction, and cpu can execute that particular line in 3 microops. That way or another, this function has way more serious performance problems, like excessive conditional branching (which is *always* expensive, for example look here: https://github.com/ceph/ceph/pull/10935) and focusing on microops of bit and arithmetic operations is a great example of wasted time because optimizing those won't bring noticeable benefits.

(Also, if you're interested, I suggest you consult Agner Fog's optimization resources (http://www.agner.org/optimize/#manuals), they contain far more and far more reliable information on current and past-gen CPUs from all three vendors).